### PR TITLE
Binary Expressions deleted by live runner make their first input symbols null.

### PR DESCRIPTION
**Pull request by mjkkirschner**
# Purpose

When a binary expression is deleted by the live runner - the binary AST is walked and all binary expressions within that subtree are assigned to null...

the change was from:
https://github.com/DynamoDS/Dynamo/pull/7282/

The behavior before this PR was that top level binary expressions were assigned to null and no AST walking was done.

I'm guessing @ke-yu was trying to support multiple assignments like:
a = 3;
b = a;
all being set to null.

but the consequence was other binary expressions like
`a && b` end up resulting in this  `a = null`

I'm opening this PR for feedback because I'm not clear on what binary expressions should actually be set to null, only assignments? Right now I am only stopping `OR` and `AND` - since that effects this particular bug, but I don't think thats the whole story.
all good.
Screen Shot 2020-03-27 at 5 41 04 PM
still good!
Screen Shot 2020-03-27 at 5 41 07 PM

### Declarations

Check these if you believe they are true

- [ ] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### CODE REVIEW

**Comment by mjkkirschner** [comment]
@aparajit-pratap please take another look, went with your third suggestion and added 2 tests. I still need to run the self serve to see if it breaks anything.

**Comment by mjkkirschner** [comment]
@aparajit-pratap unfortunately, a bunch of failures in self serve, looking into it.

**Comment by aparajit-pratap** [comment]
Curious, what did you think about making the UI logic nodes support replication?

File: src/Engine/ProtoScript/Runners/LiveRunner.cs [hunk 1]
```diff
@@ -982,7 +983,7 @@ private List<BinaryExpressionNode> BuildNullAssignments(List<AssociativeNode> as
                 }
 
                 IdentifierNode leftNode = bNode.LeftNode as IdentifierNode;
-                if (leftNode == null || leftNode.ArrayDimensions != null)
+                if (leftNode == null || leftNode.ArrayDimensions != null  || bNode.Optr != Operator.assign)
```
Comment (aparajit-pratap) [inline src/Engine/ProtoScript/Runners/LiveRunner.cs:986]: Minor, but why not have this check up here: https://github.com/DynamoDS/Dynamo/pull/10512/files#diff-2639b9fb615a6392f1215b5d9ff3c0f2R980

**Comment by mjkkirschner** [comment]
> Curious, what did you think about making the UI logic nodes support replication?

I thought that could get very difficult because the ZT versions only support two inputs and the UI ones support more than 2 - so I believe we'd end up doing binary expressions anyway - would could do both though? This fix and also the change to UI nodes?

File: src/Engine/ProtoScript/Runners/LiveRunner.cs (hunk 1 above)
Comment (mjkkirschner) [reply src/Engine/ProtoScript/Runners/LiveRunner.cs:986]: you're right, that seems like a more appropriate place, where the binary node is validated, I can move it.

File: test/Engine/ProtoTest/LiveRunnerTests/MicroFeatureTests.cs [hunk 2]
```diff
@@ -852,6 +852,93 @@ public void GraphILTest_DeletedNode01()
 
         }
 
+        [Test]
+        public void GraphILTest_DeletedBinaryExpresionDoesNotEffectReferences()
```
Comment (aparajit-pratap) [inline test/Engine/ProtoTest/LiveRunnerTests/MicroFeatureTests.cs:856]: Effectively this test is simulating the same thing that would happen in the Dynamo UI when running the graph, deleting the node, and rerunning it or is it testing something additionally?

**Comment by aparajit-pratap** [comment]
> I thought that could get very difficult because the ZT versions only support two inputs and the UI ones support more than 2 - so I believe we'd end up doing binary expressions anyway - we could do both though? This fix and also the change to UI nodes?

Probably that's something to think about for later, add to the backlog maybe? For now, I think this fix looks good.

File: test/Engine/ProtoTest/LiveRunnerTests/MicroFeatureTests.cs (hunk 2 above)
Comment (mjkkirschner) [reply test/Engine/ProtoTest/LiveRunnerTests/MicroFeatureTests.cs:856]: simulating the same thing as in the UI - I verified it failed before this change.

**Comment by mjkkirschner** [comment]
@aparajit-pratap the self serve and parallel tests all pass, gonna merge this in.

**Pull request merged**